pyinfra/api/arguments.py

```diff
@@ -71,51 +84,81 @@ def generate_env(config, value):
     "_shell_executable": {
         "description": "The shell to use. Defaults to ``sh`` (Unix) or ``cmd`` (Windows).",
         "default": lambda config: config.SHELL,
+        "type": str,
     },
     "_chdir": {
         "description": "Directory to switch to before executing the command.",
+        "type": str,
     },
     "_env": {
         "description": "Dictionary of environment variables to set.",
         "handler": generate_env,
+        "type": Mapping[str, str],
     },
     "_success_exit_codes": {
         "description": "List of exit codes to consider a success.",
         "default": lambda config: [0],
+        "type": Iterable[int],
+    },
+    "_timeout": {
+        "description": "Timeout for *each* command executed during the operation.",
+        "type": int,
+    },  # TODO not sure
+    "_get_pty": {
+        "description": "Whether to get a pseudoTTY when executing any commands.",
+        "type": bool,
+    },
+    "_stdin": {
+        "description": "String or buffer to send to the stdin of any commands.",
+        "type": str,  # TODO don't know the type for buffer
     },
-    "_timeout": "Timeout for *each* command executed during the operation.",
-    "_get_pty": "Whether to get a pseudoTTY when executing any commands.",
-    "_stdin": "String or buffer to send to the stdin of any commands.",
 }
 
 meta_kwargs = {
     # NOTE: name is the only non-_-prefixed argument
     "name": {
         "description": "Name of the operation.",
+        "type": str,
     },
     "_ignore_errors": {
         "description": "Ignore errors when executing the operation.",
         "default": lambda config: config.IGNORE_ERRORS,
+        "type": bool,
     },
-    "_precondition": "Command to execute & check before the operation commands begin.",
-    "_postcondition": "Command to execute & check after the operation commands complete.",
-    "_on_success": "Callback function to execute on success.",
-    "_on_error": "Callback function to execute on error.",
+    "_precondition": {
+        "description": "Command to execute & check before the operation commands begin.",
+        "type": str,
+    },  # TODO not sure
+    "_postcondition": {
+        "description": "Command to execute & check after the operation commands complete.",
+        "type": str,
+    },  # TODO not sure
+    "_on_success": {
+        "description": "Callback function to execute on success.",
+        "type": Callable,
```

Comment:
```suggestion
        "type": Callable[[State, Host, str], None],
```

This is probably most correct here, no return values and is passed `(state, host, operation_hash)`.